Mobile select filters fixture tmp

dannyroberts opened the pull request:
for tests. only commits starting from `pull out function for searching users in es` are new. Builds off https://github.com/dimagi/commcare-hq/pull/9336 and https://github.com/dimagi/commcare-hq/pull/9340

Probably easiest to review :blowfish:, starting with the "pull out function..." commit; full relevant diff is https://github.com/dimagi/commcare-hq/compare/mobile-select-filters-fixture...mobile-select-filters-fixture-tmp if you want a holistic view

dannyroberts committed:
add test for mobile select filters fixutres

dannyroberts committed:
fix mobile ucr fixture data with MobileSelectFilter

dannyroberts committed:
raise ucr fixutre generation errors in DEBUG

really hard to debug otherwise!

dannyroberts committed:
make mobile ucr fixture rows filterable

by adding columns to be used for Mobile Select Filters

dannyroberts committed:
add filters to mobile ucr fixtures

dannyroberts committed:
pull out function for searching users in es

get_search_mobile_workers_in_domain_es_query returning a UserES query
object

currently used in MobileWorkerListView

dannyroberts committed:
add .values_list and .values methods to ESQuery

modeled after django QuerySet’s

dannyroberts committed:
add UserChoiceProvider

dannyroberts committed:
include unrecoginized values in UserChoiceProvider.get_choices_for_values

dannyroberts committed:
include web users in UserChoiceProvider

dannyroberts committed:
use offset instead of page

this is necessary for writing a provider that daisy chains other providers
(this is what OwnerChoiceProvider will do)

dannyroberts committed:
add AbstractMultiProvider

OwnerChoiceProvider will use this

dannyroberts committed:
add GroupChoiceProvider and OwnerChoiceProvider

OwnerChoiceProvider daisy chains GroupChoiceProvider, UserChoiceProvider, and LocationChoiceProvider

dannyroberts committed:
fix ChoiceQueryContext fix attr

was being skipped if offset and page were both 0

dimagimon commented on corehq/apps/userreports/reports/filters/factory.py:
```diff
@@ -57,7 +58,7 @@ def _build_dynamic_choice_list_filter(spec):
         show_all=wrapped.show_all,
         url_generator=dynamic_choice_list_url,
-        choice_provider=FilterChoiceProviderFactory.from_spec(wrapped.get_choice_provider_spec()),
+        choice_provider=FilterChoiceProviderFactory.from_spec(wrapped.get_choice_provider_spec())(report, wrapped.slug),
```
E501 line too long (120 > 115 characters)

dannyroberts committed:
fix SuiteTest.test_case_detail_tabs_with_nodesets

which tries to make a mock CaseProvider subclass without a query_count.
Introduces a second level of ABC, ChainableChoiceProvider, that's
stricter thatn ChoiceProvider, and remove query_count from
ChoiceProvider.

That way, you only have to define query_count on ChoiceProviders that
you know are going to be chained together

dannyroberts committed:
fix UserChoiceProver.get_choices_for_known_values to include WebUsers

and move inline out of unshared helper function

dannyroberts committed:
implement GroupChoiceProvider.get_choices_for_known_values

dannyroberts committed:
have AbstractMultiProvider.get_choices_for_known_values abort when done

dannyroberts committed:
fix LocationChoiceProvider.get_choices_for_known_values

dimagimon commented on corehq/apps/users/analytics.py:
```diff
@@ -1,3 +1,5 @@
+from corehq.apps.es import UserES
+from corehq.apps.es.filters import doc_id
```
F401 'doc_id' imported but unused

dannyroberts committed:
clean up import

[ci skip]
in corehq/apps/users/analytics.py

dannyroberts commented on corehq/apps/es/es_query.py:
```diff
@@ -302,6 +302,31 @@ def remove_default_filter(self, default):
             query._default_filters.pop(default)
         return query
 
```
@esoergel I think most of this PR is better for @czue  to review, but could I get your eyes on just the changes to `corehq/apps/es/*`? Also, shoutout for ESQuery—very easy to learn how to use and extend!

esoergel replied:
Sure thing, thanks!

dannyroberts replied:
Actually, since I'm reopening as a different PR, can you comment on https://github.com/dimagi/commcare-hq/pull/9365/files#diff-6bc5251ac254d50d65a8d33297cdd7fcR302 instead?

dannyroberts commented:
Reopened rebased onto master as https://github.com/dimagi/commcare-hq/pull/9365
